File: backend-php/src/Repository/MeasurementRepository.php
<?php

namespace App\Repository;

use DateTimeImmutable;
use PDO;

final class MeasurementRepository
{
    public function __construct(private readonly PDO $pdo)
    {
    }

    public function metricExists(int $metricId): bool
    {
        $stmt = $this->pdo->prepare('SELECT 1 FROM metrics WHERE id = :metric_id');
        $stmt->execute(['metric_id' => $metricId]);

        return (bool) $stmt->fetchColumn();
    }

    public function fetchByMetricAndRange(
        int $metricId,
        DateTimeImmutable $startTime,
        DateTimeImmutable $endTime,
        int $limit
    ): array {
        $stmt = $this->pdo->prepare(
            'SELECT metric_id, timestamp, value
             FROM readings
             WHERE metric_id = :metric_id
               AND timestamp BETWEEN :start_time AND :end_time
             ORDER BY timestamp ASC
             LIMIT :row_limit'
        );

        $stmt->bindValue(':metric_id', $metricId, PDO::PARAM_INT);
        $stmt->bindValue(':start_time', $startTime->format(DATE_ATOM));
        $stmt->bindValue(':end_time', $endTime->format(DATE_ATOM));
        $stmt->bindValue(':row_limit', $limit, PDO::PARAM_INT);
        $stmt->execute();

        return $stmt->fetchAll();
    }
}

File: backend/repositories/measurement_repository.py
from datetime import datetime
from typing import Any


class MeasurementRepository:
    def __init__(self, conn: Any):
        self.conn = conn

    def metric_exists(self, metric_id: int) -> bool:
        with self.conn.cursor() as cur:
            cur.execute("SELECT 1 FROM metrics WHERE id = %s", (metric_id,))
            return cur.fetchone() is not None

    def zone_exists(self, zone_id: int) -> bool:
        with self.conn.cursor() as cur:
            cur.execute("SELECT 1 FROM zones WHERE id = %s", (zone_id,))
            return cur.fetchone() is not None

    def list_by_metric_and_range(
        self,
        metric_id: int,
        start_time: datetime,
        end_time: datetime,
        limit: int,
    ) -> list[dict[str, Any]]:
        with self.conn.cursor() as cur:
            cur.execute(
                """
                SELECT metric_id, timestamp, value
                FROM readings
                WHERE metric_id = %s
                  AND timestamp BETWEEN %s AND %s
                ORDER BY timestamp ASC
                LIMIT %s
                """,
                (metric_id, start_time, end_time, limit),
            )
            return cur.fetchall()

    def list_by_zone_and_range(
        self,
        zone_id: int,
        start_time: datetime,
        end_time: datetime,
    ) -> list[dict[str, Any]]:
        with self.conn.cursor() as cur:
            cur.execute(
                """
                SELECT r.timestamp, m.name AS metric, r.value
                FROM readings r
                JOIN metrics m ON m.id = r.metric_id
                WHERE r.zone_id = %s
                  AND r.timestamp BETWEEN %s AND %s
                ORDER BY r.timestamp ASC
                """,
                (zone_id, start_time, end_time),
            )
            return cur.fetchall()
